docs/docs/reference/configuration.md
# Configuration File

Permify offers various options for configuring your Permify API Server.

Here is the example configuration YAML file with glossary below. You can also find
this [example config file](https://github.com/Permify/permify/blob/master/example.config.yaml) in Permify repo.

***Example config.yaml file***

```yaml
# The server section specifies the HTTP and gRPC server settings,
# including whether or not TLS is enabled and the certificate and
# key file locations.
server:
  rate_limit: 100
  http:
    enabled: true
    port: 3476
    tls:
      enabled: true
      cert: /etc/letsencrypt/live/yourdomain.com/fullchain.pem
      key: /etc/letsencrypt/live/yourdomain.com/privkey.pem
  grpc:
    port: 3478
    tls:
      enabled: true
      cert: /etc/letsencrypt/live/yourdomain.com/fullchain.pem
      key: /etc/letsencrypt/live/yourdomain.com/privkey.pem

# The logger section sets the logging level for the service.
logger:
  level: info

# The profiler section enables or disables the pprof profiler and
# sets the port number for the profiler endpoint.
profiler:
  enabled: true
  port: 6060

# The authn section specifies the authentication method for the service.
authn:
  enabled: true
  method: preshared
  preshared:
    keys: [ ]

# The tracer section enables or disables distributed tracing and sets the
# exporter and endpoint for the tracing data.
tracer:
  exporter: zipkin
  endpoint: http://localhost:9411/api/v2/spans
  enabled: true

# The meter section enables or disables metrics collection and sets the
# exporter and endpoint for the collected metrics.
meter:
  exporter: otlp
  endpoint: localhost:4318
  enabled: true

# The service section sets various service-level settings, including whether
# or not to use a circuit breaker, and cache sizes for schema, permission,
# and relationship data.
service:
  circuit_breaker: false
  watch:
    enabled: false
  schema:
    cache:
      number_of_counters: 1_000
      max_cost: 10MiB
  permission:
    bulk_limit: 100
    concurrency_limit: 100
    cache:
      number_of_counters: 10_000
      max_cost: 10MiB

# The database section specifies the database engine and connection settings,
# including the URI for the database, whether or not to auto-migrate the database,
# and connection pool settings.
database:
  engine: postgres
  uri: postgres://user:password@host:5432/db_name
  auto_migrate: false
  max_open_connections: 20
  max_idle_connections: 1
  max_connection_lifetime: 300s
  max_connection_idle_time: 60s
  garbage_collection:
    enabled: true
    interval: 200h
    window: 200h
    timeout: 5m

# distributed configuration settings
distributed:
  # Indicates whether the distributed mode is enabled or not
  enabled: true

  # The address of the distributed service.
  # Using a Kubernetes DNS name suggests this service runs in a Kubernetes cluster
  # under the 'default' namespace and is named 'permify'
  address: "kubernetes:///permify.default"

  # The port on which the service is exposed
  port: "5000"

```

## Options

<details><summary>server | Server Configurations</summary>
<p>

#### Definition

Server options to run Permify. (`grpc` and `http` available for now.)

#### Structure

```
├── server
    ├── rate_limit
    ├── (`grpc` or `http`)
    │   ├── enabled
    │   ├── port
    │   └── tls
    │       ├── enabled
    │       ├── cert
    │       └── key
```

#### Glossary

| Required | Argument                  | Default | Description                                                         |
|----------|---------------------------|---------|---------------------------------------------------------------------|
| [ ]      | rate_limit                | 100     | the maximum number of requests the server should handle per second. |
| [x]      | [ server_type ]           | -       | server option type can either be `grpc` or `http`.                  |
| [ ]      | enabled (for server type) | true    | switch option for server.                                           |
| [x]      | port                      | -       | port that server run on.                                            |
| [x]      | tls                       | -       | transport layer security options.                                   |
| [ ]      | enabled (for tls)         | false   | switch option for tls                                               |
| [ ]      | cert                      | -       | tls certificate path.                                               |
| [ ]      | key                       | -       | tls key pat                                                         |

#### ENV

| Argument                  | ENV                               | Type         |
|---------------------------|-----------------------------------|--------------|
| rate_limit                | PERMIFY_RATE_LIMIT                | int          |
| grpc-port                 | PERMIFY_GRPC_PORT                 | string       |
| grpc-tls-enabled          | PERMIFY_GRPC_TLS_ENABLED          | boolean      |
| grpc-tls-key-path         | PERMIFY_GRPC_TLS_KEY_PATH         | string       |
| grpc-tls-cert-path        | PERMIFY_GRPC_TLS_CERT_PATH        | string       |
| http-enabled              | PERMIFY_HTTP_ENABLED              | boolean      |
| http-port                 | PERMIFY_HTTP_PORT                 | string       |
| http-tls-key-path         | PERMIFY_HTTP_TLS_KEY_PATH         | string       |
| http-tls-cert-path        | PERMIFY_HTTP_TLS_CERT_PATH        | string       |
| http-cors-allowed-origins | PERMIFY_HTTP_CORS_ALLOWED_ORIGINS | string array |
| http-cors-allowed-headers | PERMIFY_HTTP_CORS_ALLOWED_HEADERS | string array |

</p>
</details>

<details><summary>logger | Logging Options</summary>
<p>

#### Definition

Real time logs of authorization. Permify uses [zerolog] as a logger.

[zerolog]: https://github.com/rs/zerolog

#### Structure

```
├── logger
    ├── level
```

#### Glossary

| Required | Argument | Default | Description                                      |
|----------|----------|---------|--------------------------------------------------|
| [x]      | level    | info    | logger levels: `error`, `warn`, `info` , `debug` |
| [x]      | output   | text    | logger output: `json`, `text`                    |

#### ENV

| Argument   | ENV                | Type   |
|------------|--------------------|--------|
| log-level  | PERMIFY_LOG_LEVEL  | string |
| log-output | PERMIFY_LOG_OUTPUT | string |

</p>
</details>

<details><summary>authn | Server Authentication</summary>
<p>

#### Definition

You can choose to authenticate users to interact with Permify API.

There are 2 authentication method you can choose:

* [Pre Shared Keys](#pre-shared-keys)
* [OpenID Connect](#openid-connect)

#### Pre Shared Keys

On this method, you must provide a pre shared keys in order to identify yourself.

#### Structure

```
├── authn
|   ├── method
|   ├── enabled
|   ├── preshared
|       ├── keys
```

#### Glossary

| Required | Argument | Default | Description                                                                                                          |
|----------|----------|---------|----------------------------------------------------------------------------------------------------------------------|
| [x]      | method   | -       | Authentication method can be either `oidc` or `preshared`.                                                           |
| [ ]      | enabled  | true    | switch option authentication config                                                                                  |
| [x]      | keys     | -       | Private key/keys for server authentication. Permify does not provide this key, so it must be generated by the users. |

#### ENV

| Argument             | ENV                          | Type         |
|----------------------|------------------------------|--------------|
| authn-enabled        | PERMIFY_AUTHN_ENABLED        | boolean      |
| authn-method         | PERMIFY_AUTHN_METHOD         | string       |
| authn-preshared-keys | PERMIFY_AUTHN_PRESHARED_KEYS | string array |

#### OpenID Connect

Permify supports OpenID Connect (OIDC). OIDC provides an identity layer on top of OAuth 2.0 to address the shortcomings
of using OAuth 2.0 for establishing identity.

With this authentication method, you be able to integrate your existing Identity Provider (IDP) to validate JSON Web
Tokens (JWTs) using JSON Web Keys (JWKs). By doing so, only trusted tokens from the IDP will be accepted for
authentication.

#### Structure

```
├── authn
|   ├── method
|   ├── enabled
|   ├── oidc
|       ├── issuer
|       ├── audience
```

#### Glossary

| Required | Argument | Default | Description                                                                                                                                                                             |
|----------|----------|---------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| [x]      | method   | -       | Authentication method can be either `oidc` or `preshared`.                                                                                                                              |
| [ ]      | enabled  | false   | switch option authentication config                                                                                                                                                     |
| [x]      | audience | -       | The audience identifies the intended recipients of the token, typically the API or resource server. It ensures tokens are used only by the authorized party.                            |
| [x]      | issuer   | -       | This is the URL of the provider that is responsible for authenticating users. You will use this URL to discover information about the provider in step 1 of the authentication process. |

#### ENV

| Argument            | ENV                         | Type    |
|---------------------|-----------------------------|---------|
| authn-enabled       | PERMIFY_AUTHN_ENABLED       | boolean |
| authn-method        | PERMIFY_AUTHN_METHOD        | string  |
| authn-oidc-issuer   | PERMIFY_AUTHN_OIDC_ISSUER   | string  |
| authn-oidc-audience | PERMIFY_AUTHN_OIDC_AUDIENCE | string  |

</p>
</details>


<details><summary>tracer | Tracing Configurations</summary>
<p>

#### Definition

Permify integrated with [jaeger], [otlp], [signoz], and [zipkin] tacing tools to analyze performance and behavior of
your
authorization when using Permify.

#### Structure

```
├── tracer
|   ├── exporter
|   ├── endpoint
|   ├── enabled
|   ├── insecure
|   ├── urlpath
```

#### Glossary

| Required | Argument | Default | Description                                                                                                                   |
|----------|----------|---------|-------------------------------------------------------------------------------------------------------------------------------|
| [x]      | exporter | -       | Tracer exporter, the options are `jaeger`, `otlp`, `signoz`, and `zipkin`.                                                    |
| [x]      | endpoint | -       | export uri for tracing data.                                                                                                  |
| [ ]      | enabled  | false   | switch option for tracing.                                                                                                    |
| [ ]      | urlpath  |         | allows one to override the default URL path for otlp, used for sending traces. If unset, default ("/v1/traces") will be used. |
| [ ]      | insecure | false   | Whether to use HTTP instead of HTTPs for exporting the traces.                                                                |

#### ENV

| Argument        | ENV                     | Type    |
|-----------------|-------------------------|---------|
| tracer-enabled  | PERMIFY_TRACER_ENABLED  | boolean |
| tracer-exporter | PERMIFY_TRACER_EXPORTER | string  |
| tracer-endpoint | PERMIFY_TRACER_ENDPOINT | string  |
| tracer-urlpath  | PERMIFY_TRACER_URL_PATH | string  |
| tracer-insecure | PERMIFY_TRACER_INSECURE | boolean |

</p>
</details>

<details><summary>meter | Meter Configurations</summary>
<p>

#### Definition

Configuration for observing metrics; check count, cache check count and session information; Permify version, hostname,
os, arch.

#### Structure

```
├── meter
|   ├── exporter
|   ├── endpoint
|   ├── enabled
|   ├── insecure
|   ├── urlpath
```

#### Glossary

| Required | Argument | Default | Description                                                  |
|----------|----------|---------|--------------------------------------------------------------|
| [x]      | exporter | -       | [otlp](https://opentelemetry.io/docs/collector/) is default. |
| [x]      | endpoint | -       | export uri for metric observation                            |
| [ ]      | enabled  | true    | switch option for meter tracing.                             |

#### ENV

| Argument       | ENV                    | Type    |
|----------------|------------------------|---------|
| meter-enabled  | PERMIFY_METER_ENABLED  | boolean |
| meter-exporter | PERMIFY_METER_EXPORTER | string  |
| meter-endpoint | PERMIFY_METER_ENDPOINT | string  |
| meter-urlpath  | PERMIFY_METER_URL_PATH | string  |
| meter-insecure | PERMIFY_METER_INSECURE | boolean |

</p>
</details>

<details><summary>database | Database Configurations</summary>
<p>

#### Definition

Configurations for the database that points out where your want to store your authorization data (relation tuples,
audits, decision logs, authorization model)

#### Structure

```
├── database
|   ├── engine
|   ├── uri
|   ├── auto_migrate
|   ├── max_open_connections
|   ├── max_idle_connections
|   ├── max_connection_lifetime
|   ├── max_connection_idle_time
|   ├──garbage_collection
|       ├──enable: true
|       ├──interval: 3m
|       ├──timeout: 3m
|       ├──window: 720h
```

#### Glossary

| Required | Argument                        | Default | Description                                                                                                       |
|----------|---------------------------------|---------|-------------------------------------------------------------------------------------------------------------------|
| [x]      | engine                          | memory  | Data source. Permify supports **PostgreSQL**(`'postgres'`) for now. Contact with us for your preferred database.  |
| [x]      | uri                             | -       | Uri of your data source.                                                                                          |
| [ ]      | auto_migrate                    | true    | When its configured as false migrating flow won't work.                                                           |                                           
| [ ]      | max_open_connections            | 20      | Configuration parameter determines the maximum number of concurrent connections to the database that are allowed. |
| [ ]      | max_idle_connections            | 1       | Determines the maximum number of idle connections that can be held in the connection pool.                        |
| [ ]      | max_connection_lifetime         | 300s    | Determines the maximum lifetime of a connection in seconds.                                                       |                 
| [ ]      | max_connection_idle_time        | 60s     | Determines the maximum time in seconds that a connection can remain idle before it is closed.                     |                
| [ ]      | enable (for garbage collection) | false   | Switch option for garbage collection.                                                                             |               
| [ ]      | interval                        | 3m      | Determines the run period of a Garbage Collection operation.                                                      |              
| [ ]      | timeout                         | 3m      | Sets the duration of the Garbage Collection timeout.                                                              |             
| [ ]      | window                          | 720h    | Determines how much backward cleaning the Garbage Collection process will perform.                                |                     

#### ENV

| Argument                             | ENV                                          | Type     |
|--------------------------------------|----------------------------------------------|----------|
| database-engine                      | PERMIFY_DATABASE_ENGINE                      | string   |
| database-uri                         | PERMIFY_DATABASE_URI                         | string   |
| database-auto-migrate                | PERMIFY_DATABASE_AUTO_MIGRATE                | boolean  |
| database-max-open-connections        | PERMIFY_DATABASE_MAX_OPEN_CONNECTIONS        | int      |
| database-max-idle-connections        | PERMIFY_DATABASE_MAX_IDLE_CONNECTIONS        | int      |
| database-max-connection-lifetime     | PERMIFY_DATABASE_MAX_CONNECTION_LIFETIME     | duration |
| database-max-connection-idle-time    | PERMIFY_DATABASE_MAX_CONNECTION_IDLE_TIME    | duration |
| database-garbage-collection-enabled  | PERMIFY_DATABASE_GARBAGE_ENABLED             | boolean  |
| database-garbage-collection-interval | PERMIFY_DATABASE_GARBAGE_COLLECTION_INTERVAL | duration |
| database-garbage-collection-timeout  | PERMIFY_DATABASE_GARBAGE_COLLECTION_TIMEOUT  | duration |
| database-garbage-collection-window   | PERMIFY_DATABASE_GARBAGE_COLLECTION_WINDOW   | duration |

</p>
</details>

<details><summary>service | Service Configurations</summary>
<p>

#### Definition

Configurations for the permify service and how it should behave. You can configure the circuit breaker pattern,
configuration watcher, and service specific options for permission and schema services (rate limiting, concurrency
limiting, cache size).

#### Structure

```
├── service
|   ├── circuit_breaker
|   ├── watch:
|   |   ├── enabled
|   ├── schema:
|   |   ├── cache:
|   |   |   ├── number_of_counters
|   |   |   ├── max_cost
|   |   permission:
|   |   |   ├── bulk_limit
|   |   |   ├── concurrency_limit
|   |   |   ├── cache:
|   |   |   |   ├── number_of_counters
|   |   |   |   ├── max_cost
```

#### Glossary

| Required | Argument                        | Default | Description                                       |
|----------|---------------------------------|---------|---------------------------------------------------|
| [ ]      | circuit_breaker                 | false   | switch option to use the circuit breaker pattern. |
| [ ]      | watch                           | false   | switch option for configuration watcher.          |
| [ ]      | schema.cache.number_of_counters | 1_000   | number of counters for schema service.            |
| [ ]      | schema.cache.max_cost           | 10MiB   | max cost for schema cache.                        |
| [ ]      | permission.bulk_limit           | 100     | bulk operations limit for permission service.     |
| [ ]      | permission.concurrency_limit    | 100     | concurrency limit for permission service.         |
| [ ]      | permission.cache.max_cost       | 10MiB   | max cost for permission service.                  |

#### ENV

| Argument                                | ENV                                             | Type    |
|-----------------------------------------|-------------------------------------------------|---------|
| service-circuit-breaker                 | PERMIFY_SERVICE_CIRCUIT_BREAKER                 | boolean |
| service-watch-enabled                   | PERMIFY_SERVICE_WATCH_ENABLED                   | boolean |
| service-schema-cache-number-of-counters | PERMIFY_SERVICE_SCHEMA_CACHE_NUMBER_OF_COUNTERS | int     |
| service-schema-cache-max-cost           | PERMIFY_SERVICE_SCHEMA_CACHE_MAX_COST           | int     |
| service-permission-bulk-limit           | PERMIFY_SERVICE_PERMISSION_BULK_LIMIT           | int     |
| service-permission-concurrency-limit    | PERMIFY_SERVICE_PERMISSION_CONCURRENCY_LIMIT    | int     |
| service-permission-cache-max-cost       | PERMIFY_SERVICE_PERMISSION_CACHE_MAX_COST       | int     |

</p>
</details>

<details><summary>profiler | Performance Profiler Configurations</summary>
<p>

#### Definition

pprof is a performance profiler for Go programs. It allows developers to analyze and understand the performance
characteristics of their code by generating detailed profiles of program execution

#### Structure

```
├── profiler
|   ├── enabled
|   ├── port
```

#### Glossary

| Required | Argument | Default | Description                                   |
|----------|----------|---------|-----------------------------------------------|
| [ ]      | enabled  | true    | switch option for profiler.                   |
| [x]      | port     | -       | port that profiler runs on *(default: 6060)*. |

#### ENV

| Argument         | ENV                      | Type    |
|------------------|--------------------------|---------|
| profiler-enabled | PERMIFY_PROFILER_ENABLED | boolean |
| profiler-port    | PERMIFY_PROFILER_PORT    | string  |

</p>
</details>

<details><summary>Distributed | Consistent hashing Configurations</summary>
<p>

#### Definition

A consistent hashing ring ensures data distribution that minimizes reorganization when nodes are added or removed,
improving scalability and performance in distributed systems."

#### Structure

```
├── distributed
|   ├── enabled
|   ├── address
|   ├── port
```

#### Glossary

| Required | Argument | Default | Description                          |
|----------|----------|---------|--------------------------------------|
| [x]      | enabled  | false   | switch option for distributed.       |
| []       | address  | -       | address of the distributed service   |
| []       | port     | 5000    | port on which the service is exposed |

#### ENV

| Argument            | ENV                         | Type    |
|---------------------|-----------------------------|---------|
| distributed-enabled | PERMIFY_DISTRIBUTED_ENABLED | boolean |
| distributed-address | PERMIFY_DISTRIBUTED_ADDRESS | string  |
| distributed-port    | PERMIFY_DISTRIBUTED_PORT    | string  |

</p>
</details>


## Testing Config

You can use the `permify config` command to view the latest state of your configuration. The `Value` part represents the configuration that will be executed, and the `Source` part indicates where this value comes from. There are three options for the "Source" part: (`ENV`, `FILE`, `FLAG`)

![testing-configuration](https://github.com/Permify/permify/assets/39353278/1b376e81-1fb1-4ccb-b379-a66226d272a2)

[jaeger]: https://www.jaegertracing.io/

[otlp]: https://opentelemetry.io/

[zipkin]: https://zipkin.io/

[signoz]: https://signoz.io/


internal/config/config.go
package config

import (
	"fmt"
	"path/filepath"
	"time"

	"github.com/pkg/errors"

	"github.com/spf13/viper"
)

type (
	// Config is the main configuration structure containing various sections for different aspects of the application.
	Config struct {
		AccountID   string                       `mapstructure:"account_id"`
		Server      `mapstructure:"server"`      // Server configuration for both HTTP and gRPC
		Log         `mapstructure:"logger"`      // Logging configuration
		Profiler    `mapstructure:"profiler"`    // Profiler configuration
		Authn       `mapstructure:"authn"`       // Authentication configuration
		Tracer      `mapstructure:"tracer"`      // Tracing configuration
		Meter       `mapstructure:"meter"`       // Metrics configuration
		Service     `mapstructure:"service"`     // Service configuration
		Database    `mapstructure:"database"`    // Database configuration
		Distributed `mapstructure:"distributed"` // Distributed configuration
	}

	// Server contains the configurations for both HTTP and gRPC servers.
	Server struct {
		HTTP      `mapstructure:"http"` // HTTP server configuration
		GRPC      `mapstructure:"grpc"` // gRPC server configuration
		RateLimit int64                 `mapstructure:"rate_limit"` // Rate limit configuration
	}

	// HTTP contains configuration for the HTTP server.
	HTTP struct {
		Enabled            bool      `mapstructure:"enabled"`              // Whether the HTTP server is enabled
		Port               string    `mapstructure:"port"`                 // Port for the HTTP server
		TLSConfig          TLSConfig `mapstructure:"tls"`                  // TLS configuration for the HTTP server
		CORSAllowedOrigins []string  `mapstructure:"cors_allowed_origins"` // List of allowed origins for CORS
		CORSAllowedHeaders []string  `mapstructure:"cors_allowed_headers"` // List of allowed headers for CORS
	}

	// GRPC contains configuration for the gRPC server.
	GRPC struct {
		Port      string    `mapstructure:"port"` // Port for the gRPC server
		TLSConfig TLSConfig `mapstructure:"tls"`  // TLS configuration for the gRPC server
	}

	// TLSConfig contains configuration for TLS.
	TLSConfig struct {
		Enabled  bool   `mapstructure:"enabled"` // Whether TLS is enabled
		CertPath string `mapstructure:"cert"`    // Path to the certificate file
		KeyPath  string `mapstructure:"key"`     // Path to the key file
	}

	// Authn contains configuration for authentication.
	Authn struct {
		Enabled   bool      `mapstructure:"enabled"`   // Whether authentication is enabled
		Method    string    `mapstructure:"method"`    // The authentication method to be used
		Preshared Preshared `mapstructure:"preshared"` // Configuration for preshared key authentication
		Oidc      Oidc      `mapstructure:"oidc"`      // Configuration for OIDC authentication
	}

	// Preshared contains configuration for preshared key authentication.
	Preshared struct {
		Keys []string `mapstructure:"keys"` // List of preshared keys
	}

	// Oidc contains configuration for OIDC authentication.
	Oidc struct {
		Issuer   string `mapstructure:"issuer"`   // OIDC issuer URL
		Audience string `mapstructure:"audience"` // OIDC client ID
	}

	// Profiler contains configuration for the profiler.
	Profiler struct {
		Enabled bool   `mapstructure:"enabled"` // Whether the profiler is enabled
		Port    string `mapstructure:"port"`    // Port for the profiler
	}

	// Log contains configuration for logging.
	Log struct {
		Level  string `mapstructure:"level"`  // Logging level
		Output string `mapstructure:"output"` // Logging output format, e.g., text, json
	}

	// Tracer contains configuration for distributed tracing.
	Tracer struct {
		Enabled  bool   `mapstructure:"enabled"`  // Whether tracing collection is enabled
		Exporter string `mapstructure:"exporter"` // Exporter for tracing data
		Endpoint string `mapstructure:"endpoint"` // Endpoint for the tracing exporter
		Insecure bool   `mapstructure:"insecure"` // Connect to the collector using the HTTP scheme, instead of HTTPS.
		URLPath  string `mapstructure:"path"`     // Path for the tracing exporter, if not defined /v1/trace will be used
	}

	// Meter contains configuration for metrics collection and reporting.
	Meter struct {
		Enabled  bool   `mapstructure:"enabled"`  // Whether metrics collection is enabled
		Exporter string `mapstructure:"exporter"` // Exporter for metrics data
		Endpoint string `mapstructure:"endpoint"` // Endpoint for the metrics exporter
		Insecure bool   `mapstructure:"insecure"` // Connect to the collector using the HTTP, GRPC scheme, instead of HTTPS, GRPC secure.
		URLPath  string `mapstructure:"path"`     // Path for the metrics exporter, if not defined /v1/metrics will be used, will be ignored in case of GRPC
		Grpc     bool   `mapstructure:"grpc"`     // Connect to GRPC exporter
	}

	// Service contains configuration for various service-level features.
	Service struct {
		CircuitBreaker bool       `mapstructure:"circuit_breaker"` // Whether to enable the circuit breaker pattern
		Watch          Watch      `mapstructure:"watch"`           // Watch service configuration
		Schema         Schema     `mapstructure:"schema"`          // Schema service configuration
		Permission     Permission `mapstructure:"permission"`      // Permission service configuration
		Data           Data       `mapstructure:"data"`            // Data service configuration
	}

	// Watch contains configuration for the watch service.
	Watch struct {
		Enabled bool `mapstructure:"enabled"`
	}

	// Schema contains configuration for the schema service.
	Schema struct {
		Cache Cache `mapstructure:"cache"` // Cache configuration for the schema service
	}

	// Permission contains configuration for the permission service.
	Permission struct {
		BulkLimit        int   `mapstructure:"bulk_limit"`        // Limit for bulk operations
		ConcurrencyLimit int   `mapstructure:"concurrency_limit"` // Limit for concurrent operations
		Cache            Cache `mapstructure:"cache"`             // Cache configuration for the permission service
	}

	// Data is a placeholder struct for the data service configuration.
	Data struct{}

	// Cache contains configuration for caching.
	Cache struct {
		NumberOfCounters int64  `mapstructure:"number_of_counters"` // Number of counters for the cache
		MaxCost          string `mapstructure:"max_cost"`           // Maximum cost for the cache
	}

	// Database contains configuration for the database.
	Database struct {
		Engine                string            `mapstructure:"engine"`                  // Database engine type (e.g., "postgres" or "memory")
		URI                   string            `mapstructure:"uri"`                     // Database connection URI
		AutoMigrate           bool              `mapstructure:"auto_migrate"`            // Whether to enable automatic migration
		MaxOpenConnections    int               `mapstructure:"max_open_connections"`    // Maximum number of open connections to the database
		MaxIdleConnections    int               `mapstructure:"max_idle_connections"`    // Maximum number of idle connections to the database
		MaxConnectionLifetime time.Duration     `mapstructure:"max_connection_lifetime"` // Maximum duration a connection can be reused
		MaxConnectionIdleTime time.Duration     `mapstructure:"max_connection_idle_time"`
		GarbageCollection     GarbageCollection `mapstructure:"garbage_collection"`
	}

	GarbageCollection struct {
		Enabled  bool          `mapstructure:"enabled"`
		Interval time.Duration `mapstructure:"interval"`
		Timeout  time.Duration `mapstructure:"timeout"`
		Window   time.Duration `mapstructure:"window"`
	}

	Distributed struct {
		Enabled bool   `mapstructure:"enabled"`
		Address string `mapstructure:"address"`
		Port    string `mapstructure:"port"`
	}
)

// NewConfig initializes and returns a new Config object by reading and unmarshalling
// the configuration file from the given path. It falls back to the DefaultConfig if the
// file is not found. If there's an error during the process, it returns the error.
func NewConfig() (*Config, error) {
	// Start with the default configuration values
	cfg := DefaultConfig()

	// Set the name and type of the config file to be read
	viper.SetConfigName("config")
	viper.SetConfigType("yaml")

	// Add the path where the config file is located
	viper.AddConfigPath("./config")

	// Read the config file
	err := viper.ReadInConfig()
	// If there's an error during reading the config file
	if err != nil {
		// Check if the error is because of the config file not being found
		if ok := errors.As(err, &viper.ConfigFileNotFoundError{}); !ok {
			// If it's not a "file not found" error, return the error with a message
			return nil, fmt.Errorf("failed to load server config: %w", err)
		}
		// If it's a "file not found" error, the code will continue and use the default config
	}

	// Unmarshal the configuration data into the Config struct
	if err = viper.Unmarshal(cfg); err != nil {
		// If there's an error during unmarshalling, return the error with a message
		return nil, fmt.Errorf("failed to unmarshal server config: %w", err)
	}

	// Return the populated Config object
	return cfg, nil
}

// NewConfigWithFile initializes and returns a new Config object by reading and unmarshalling
// the configuration file from the given path. It falls back to the DefaultConfig if the
// file is not found. If there's an error during the process, it returns the error.
func NewConfigWithFile(dir string) (*Config, error) {
	// Start with the default configuration values
	cfg := DefaultConfig()

	viper.SetConfigFile(dir)

	err := isYAML(dir)
	if err != nil {
		return nil, err
	}

	// Read the config file
	err = viper.ReadInConfig()
	// If there's an error during reading the config file
	if err != nil {
		// Check if the error is because of the config file not being found
		if ok := errors.As(err, &viper.ConfigFileNotFoundError{}); !ok {
			// If it's not a "file not found" error, return the error with a message
			return nil, fmt.Errorf("failed to load server config: %w", err)
		}
		if ok := errors.As(err, &viper.ConfigMarshalError{}); !ok {
			// If it's not a "file not found" error, return the error with a message
			return nil, fmt.Errorf("failed to load server config: %w", err)
		}
		// If it's a "file not found" error, the code will continue and use the default config
	}

	// Unmarshal the configuration data into the Config struct
	if err = viper.Unmarshal(cfg); err != nil {
		// If there's an error during unmarshalling, return the error with a message
		return nil, fmt.Errorf("failed to unmarshal server config: %w", err)
	}

	// Return the populated Config object
	return cfg, nil
}

// DefaultConfig - Creates default config.
func DefaultConfig() *Config {
	return &Config{
		AccountID: "",
		Server: Server{
			HTTP: HTTP{
				Enabled: true,
				Port:    "3476",
				TLSConfig: TLSConfig{
					Enabled: false,
				},
				CORSAllowedOrigins: []string{"*"},
				CORSAllowedHeaders: []string{"*"},
			},
			GRPC: GRPC{
				Port: "3478",
				TLSConfig: TLSConfig{
					Enabled: false,
				},
			},
			RateLimit: 10_000,
		},
		Profiler: Profiler{
			Enabled: false,
		},
		Log: Log{
			Level: "info",
		},
		Tracer: Tracer{
			Enabled: false,
		},
		Meter: Meter{
			Enabled:  true,
			Exporter: "otlp",
			Endpoint: "telemetry.permify.co",
			Grpc:     false,
		},
		Service: Service{
			CircuitBreaker: false,
			Watch: Watch{
				Enabled: false,
			},
			Schema: Schema{
				Cache: Cache{
					NumberOfCounters: 1_000,
					MaxCost:          "10MiB",
				},
			},
			Permission: Permission{
				BulkLimit:        100,
				ConcurrencyLimit: 100,
				Cache: Cache{
					NumberOfCounters: 10_000,
					MaxCost:          "10MiB",
				},
			},
			Data: Data{},
		},
		Authn: Authn{
			Enabled:   false,
			Preshared: Preshared{},
			Oidc:      Oidc{},
		},
		Database: Database{
			Engine:      "memory",
			AutoMigrate: true,
			GarbageCollection: GarbageCollection{
				Enabled: false,
			},
		},
		Distributed: Distributed{
			Enabled: false,
			Port:    "5000",
		},
	}
}

func isYAML(file string) error {
	ext := filepath.Ext(file)
	if ext != ".yaml" {
		return errors.New("file is not yaml")
	}
	return nil
}
